Kafka 3368

andrasbeni opened the pull request:
Added Message set / Record set documentation to protocol.html. Also added short description of new types used in 0.11

Commit:
KAFKA-3368: Add the Message/Record set protocol to the protocol docs

I decided to add the documentation manually instaed of generating it.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5678/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5664/
Test PASSed (JDK 8 and Scala 2.12).

andrasbeni commented:
@ijuma I just realized you added this PR to 1.1.0 milestone. Is there anything I can do to ease merging?

andrasbeni commented:
@ijuma, @hachikuji , @mjsax I lost track of current development, so I don't know if the contents of the PR are still up to date. Should I refresh it? Or simply close it as it is uninteresting to the community?

hachikuji commented:
@andrasbeni Sorry for the delay. Been meaning to look at this, just haven't had the time. I know we already have some documentation for the message format in docs/implementation.html. I think you may have some additional detail here (for example, I don't see the old message format documented in the current doc), so does it make sense to try to update the docs there?

Commit:
Move MessageSet doc to implementation.html

Change-Id: I286c4c4fbc0b79317c0b74e787ec8c8bb64c202f

Commit:
Minor fix in html

Change-Id: I7f2c9b6c6d1dc9d097553244df52f760499175aa

andrasbeni commented on docs/protocol.html:
```diff
@@ -185,17 +185,16 @@ <h5><a id="protocol_common" href="#protocol_common">Common Request and Response
-
-<p>A description of the message set format can be found <a href="https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-Messagesets">here</a>. (KAFKA-3368)</p>
+<h5><a id="protocol_recordbatch" href="#protocol_recordbatch">Record Batch</a></h5>
+<p>A description of the record batch format can be found <a href="documentation/#recordbatch">here</a>.</p>
```
I am not sure this href will work, but I don't know how to test.

hachikuji replied:
We're all over the place with our hrefs, so I can understand the confusion. I think we're just missing a slash: `/documentation/#recordbatch`.

andrasbeni commented:
@hachikuji How about this?

hachikuji commented on docs/implementation.html:
```diff
@@ -78,7 +78,49 @@ <h5><a id="controlbatch" href="#controlbatch">5.3.1.1 Control Batches</a></h5>
     <p>The schema for the value of a control record is dependent on the type. The value is opaque to clients.</p>
 
-	<h4><a id="record" href="#record">5.3.2 Record</a></h4>
+    <h5><a id="messageset" href="#messageset">5.3.1.2 Old Message Format</a></h5>
```
Would it make more sense to describe the new format first?

hachikuji commented on docs/implementation.html:
```diff
@@ -78,7 +78,49 @@ <h5><a id="controlbatch" href="#controlbatch">5.3.1.1 Control Batches</a></h5>
+    </pre></p>
+    <p>
+        When the producer is sending non compressed messages, it can set the offsets to anything.
+        When the producer is sending compressed messages, to avoid server side re-compression, each compressed message should have offset starting from 0 and increasing by one for each inner message in the compressed message.
```
I think we should emphasize that this capability is only supported in v1. For v0 messages, the inner offsets are not relative. For the purpose of the producer, it does not really matter since offsets always start at 0, but for the consumer the distinction is important.

hachikuji commented on docs/implementation.html:
```diff
@@ -78,7 +78,49 @@ <h5><a id="controlbatch" href="#controlbatch">5.3.1.1 Control Batches</a></h5>
+        When the producer is sending compressed messages, to avoid server side re-compression, each compressed message should have offset starting from 0 and increasing by one for each inner message in the compressed message.
+    </p>
+    <p>
+        In versions until Kafka 0.10 the magic byte value was 0. Value 1 i.e. message format version 1 was introduced with timestamp support in version 0.10 .
```
nit: "versions prior to 0.10 ..."

hachikuji commented on docs/implementation.html:
```diff
@@ -78,7 +78,49 @@ <h5><a id="controlbatch" href="#controlbatch">5.3.1.1 Control Batches</a></h5>
+    <p>
+        In versions until Kafka 0.10 the magic byte value was 0. Value 1 i.e. message format version 1 was introduced with timestamp support in version 0.10 .
+        In version 1 the fourth lowest bit of attributes byte represents the timestamp type. 0 stands for CreateTime and 1 stands for LogAppendTime.
+        The producer should always set this bit to 0. Similarly to version 2 above, the lowest bits of attributes mean compression type and the highest bits must be set to 0.
```
Maybe we can mention that the broker will overwrite the timestamp if the configured timestamp type for a topic is log append time?

hachikuji commented on docs/implementation.html:
```diff
@@ -78,7 +78,49 @@ <h5><a id="controlbatch" href="#controlbatch">5.3.1.1 Control Batches</a></h5>
+        In version 1 the fourth lowest bit of attributes byte represents the timestamp type. 0 stands for CreateTime and 1 stands for LogAppendTime.
+        The producer should always set this bit to 0. Similarly to version 2 above, the lowest bits of attributes mean compression type and the highest bits must be set to 0.
+    </p>
+    <p>In message format versions 0 and 1 Kafka supports recursive messages in which case the value field may itself contain a message set.</p>
```
We should probably mention that this is only for compressed message sets. Also, maybe this can go above the comment above about compression which refers to "inner messages."

hachikuji reviewed (commented):
Thanks for the patch! Left a few comments.

Commit:
Addressing Jason's comments:
 - Moving the section after Record format specs
 - Detailing offset overwrites
 - Detailing timestamp types and timestamp overwrites
 - Clarifying recursive messages must be compressed
 - Language use

Change-Id: I9a6271ba98722e9c0ca972022c67b37cab718e42

andrasbeni commented:
@hachikuji,
I believe I addressed your concerns with the latest commit. Can you please take another look at this patch?

hachikuji commented on docs/implementation.html:
```diff
@@ -92,7 +92,7 @@ <h4><a id="record" href="#record">5.3.2 Record</a></h4>
 		Headers => [Header]
 	</pre></p>
-	<h5><a id="recordheader" href="#recordheader">5.4.2.1 Record Header</a></h5>
+	<h5><a id="recordheader" href="#recordheader">5.3.2.1 Record Header</a></h5>
```
Good catch 👍

hachikuji commented on docs/implementation.html:
```diff
@@ -15,7 +15,7 @@
  limitations under the License.
 -->
 
-<script id="implementation-template" type="text/x-handlebars-template">
```
Was this intentional? We probably need to revert.

hachikuji commented on docs/implementation.html:
```diff
@@ -102,6 +102,56 @@ <h5><a id="recordheader" href="#recordheader">5.4.2.1 Record Header</a></h5>
 
+    <h4><a id="messageset" href="#messageset">5.3.3 Old Message Format</a></h4>
+    <p>
+        Prior to Kafka 0.11 messages were transferred and stored in a format called MessageSets. In MessageSets each message has its own metadata. Note that although MessageSets are formally arrays,
```
nit: do we need to pluralize `MessageSet`?

hachikuji commented on docs/implementation.html:
```diff
@@ -102,6 +102,56 @@ <h5><a id="recordheader" href="#recordheader">5.4.2.1 Record Header</a></h5>
+        When the producer is sending non compressed messages, it can set the offsets to anything.
+    </p>
+
+    <p>The crc field contains the CRC32 (and not CRC-32C) of the message bytes.</p>
```
Maybe worthwhile emphasizing that the CRC covers all fields from the magic to the message value?

hachikuji commented on docs/implementation.html:
```diff
@@ -102,6 +102,56 @@ <h5><a id="recordheader" href="#recordheader">5.4.2.1 Record Header</a></h5>
+            value => BYTES
+    </pre></p>
+    <p>
+        In versions prior to Kafka 0.10 the magic byte value was 0. Message format version 1 was introduced with timestamp support in version 0.10 .
```
It's a little annoying to switch between "magic" and "message format version." Maybe we could rephrase the first sentence like this:
> In versions prior to 0.10, the only supported message format version (which is indicated in the magic value) was 0.

hachikuji commented on docs/implementation.html:
```diff
@@ -102,6 +102,56 @@ <h5><a id="recordheader" href="#recordheader">5.4.2.1 Record Header</a></h5>
+    <p>
+        In versions prior to Kafka 0.10 the magic byte value was 0. Message format version 1 was introduced with timestamp support in version 0.10 .
+    <ul>
+        <li>Similarly to version 2 above, the lowest bits of attributes mean compression type.</li>
```
Rather than describing the attributes in prose, could we represent them in the schema as we did above for the old message format:
```
 		attributes: int16
 			bit 0~2:
 				0: no compression
 				1: gzip
 				2: snappy
 				3: lz4
 			bit 3: timestampType
 			bit 4: isTransactional (0 means not transactional)
 			bit 5: isControlBatch (0 means not a control batch)
 			bit 6~15: unused
```

The explanation of the use of the timestamp type is useful though.

hachikuji reviewed (commented):
Thanks for the updates. Just a few more minor comments.

Commit:
Minor changes to clarify things and remove annoyances

Change-Id: I8d7ae5ae3bb33ec5a6908eecbcbad0f1ab4f08e6

Commit:
Revert another unintentional change

Change-Id: I97b46a45ed1d6753dda2092d0593a3ea02f5978e

hachikuji committed:
A few tweaks to try and clarify recursive message sets

hachikuji commented:
@andrasbeni Thanks for the update. I added a commit with a few tweaks. If it looks good to you, I will go ahead and merge.

andrasbeni commented:
@hachikuji It looks good to me.
Thank you for the review.  
Do you think there is any point in working on other subtasks of KAFKA-3360?

hachikuji commented:
@andrasbeni I think there is definitely value in finishing the work. At the moment, the protocol page doesn't appear to be linked into the main docs (at least I can't find it on kafka.apache.org). Once we fix that, we can consider moving the message format docs there. And once we have automatic generation of API schemas, then it would be a really great addition to the current docs.

hachikuji reviewed (approved):
LGTM. Thanks for the patch (and your patience with review)!
